plugins/sep/colorconfig/CustomColorOperations.cc

```diff
@@ -0,0 +1,162 @@
+    int                                           stride = spp * (tile->width - area.getWidth());
+    Scroom::Bitmap::SampleIterator<const uint8_t> si(tile->data.get(), 0, bps);
+    si += offset;
+
+    std::map<std::string, size_t> sums = {};
+
+    // Initialize sums of all colors to 0
+    for (auto color : colors){
+        sums[color->getName()] = 0;
+    }
+
+    for(int y = area.getTop(); y < area.getBottom(); y++)
+    {
+        for(int x = area.getLeft(); x < area.getRight(); x++)
+        {
+            for (auto color: colors){
+                sums[color->getName()] += *si++;
+            }
+        }
+        si += stride;
+    }
+
+    // Copy map to vector of pairs
+    PipetteColor result(sums.size());
+    copy(sums.begin(), sums.end(), result.begin());
+
+    return result;
+}
+
+
+
+
+OperationsCustomColors::OperationsCustomColors() : PipetteCommonOperationsCustomColor(8){}
+
+PipetteCommonOperationsCustomColor::Ptr OperationsCustomColors::create() { return PipetteCommonOperationsCustomColor::Ptr(new OperationsCustomColors()); }
+
+Scroom::Utils::Stuff OperationsCustomColors::cache(const ConstTile::Ptr tile) {
+
+    // Allocate the space for the cache - stride is the height of one row
+    const int stride = cairo_format_stride_for_width(CAIRO_FORMAT_ARGB32, tile->width);
+    boost::shared_ptr<uint8_t> data = shared_malloc(static_cast<size_t>(stride * tile->height));
+
+    // Row is a pointer to a row of pixels (destination)
+    auto *row = reinterpret_cast<uint32_t *>(data.get());
+    // Cur is a pointer to the start of the row in the tile (source)
+    const uint8_t *cur = tile->data.get();
+
+    for (int i = 0; i < spp * tile->height * tile->width; i += spp) {
+        // Convert custom colors to CMKY and then to ARGB, because cairo doesn't know how to render CMYK.
+        int32_t C = 0;
+        int32_t M = 0;
+        int32_t Y = 0;
+        int32_t K = 0;
+        for (int j = 0; j < spp; j++){
+            auto color = colors.at(j);
+            C += color->cMultiplier * static_cast<float>(cur[i + j]);
+            M += color->mMultiplier * static_cast<float>(cur[i + j]);
+            Y += color->yMultiplier * static_cast<float>(cur[i + j]);
+            K += color->kMultiplier * static_cast<float>(cur[i + j]);
+        }
+
+        auto C_i = static_cast<uint8_t>(255 - CustomColorHelpers::toUint8(C));
+        auto M_i = static_cast<uint8_t>(255 - CustomColorHelpers::toUint8(M));
+        auto Y_i = static_cast<uint8_t>(255 - CustomColorHelpers::toUint8(Y));
+        auto K_i = static_cast<uint8_t>(255 - CustomColorHelpers::toUint8(K));
+
+        uint32_t R = static_cast<uint8_t>((C_i * K_i) / 255);
+        uint32_t G = static_cast<uint8_t>((M_i * K_i) / 255);
+        uint32_t B = static_cast<uint8_t>((Y_i * K_i) / 255);
+
+        // Write 255 as alpha (fully opaque)
+        int target  = i * 4 / spp ; // Scale the target to the 4 channel target row, from the n channel source row
```

Comment:
I don't think that's correct. This means that if I have a Cyan value of 255, and it contributes 1, then if I have more than 4 channels, I get less Cyan...

Reply:
It is correct. The i variable iterates over a source row with n samples per pixel. The target row has 4 spp, as it is ARGB. So all the the target index is not the same as i, but needs to be scaled to write to the correct pixel in the ciaro buffer.